semgrep/semgrep/core_runner.py

```diff
@@ -441,11 +457,19 @@ def _run_rules(
             ):
                 debug_tqdm_write(f"Running rule {rule._raw.get('id')}")
                 rule_matches, debugging_steps, errors = self._run_rule(
-                    rule, target_manager, semgrep_core_ast_cache_dir
+                    rule, target_manager, semgrep_core_ast_cache_dir, max_timeout_files
                 )
                 findings_by_rule[rule] = rule_matches
                 debugging_steps_by_rule[rule] = debugging_steps
                 all_errors.extend(errors)
+                for err in errors:
+                    if isinstance(err, MatchTimeoutError):
+                        file_timeouts[err.path] += 1
+                        if (
+                            self._timeout_threshold != 0
+                            and file_timeouts[err.path] == self._timeout_threshold
```

Comment:
We can be slightly more defensive here by using `>=`, which could prevent an infinite loop in the future.